### ISSUE TITLE
Possible "ohCrap" Memory Leak!

### ISSUE THREAD

**Issue by LeonanCarvalho** [angular-ui/angular-google-maps#1331]
Hello everyone,

I'm using your maps since version 2.0.x, We are developing a very large SPA and maps is an essential resource.

Few days ago I began the deploy phase, and I got some problems with memory leaks. To to be realistic I never mind this.

Researching a lot about that this last week I have found a memory leak related to a "ohCrap" function (that should be called "Memory Leak Factory" lol :smiling_imp:) inside  uiGmapIsReady service.

This function call itself (recursively) a intensive and massive times until the browser crash(Chrome, mozilla, IE, whatever).

```javascript
 ohCrap = function() {
            return $timeout(function() {
              if (ctr !== expect) {
                return ohCrap(); //Oh crap, a lot of crap!
              } else {
                return d.resolve(promises());
              }
            });//No timeout, a bad idea!
          };
          ohCrap();//Lets start this crap!
```

I've recorded the timeline in 2 cases, 1st using normal ohCrap function, 2nd with ohCrap but setting timeout to 1000,  both cases there is no pagination just a idle visualization.

take a look:

# ohCrap Memory Leak: 
![ohcrap](https://cloud.githubusercontent.com/assets/3593706/7817171/69f4b666-03a9-11e5-86d2-636fd201c778.PNG)

# ohCrap using timeout 1000 (its not solves but reduce leak):

![timedoutohcrap](https://cloud.githubusercontent.com/assets/3593706/7817122/fc4829ae-03a8-11e5-8028-bf177bc81b5e.png)

Note that still have a growing memory but its stabilizes JS Heap SIze below compared to without timeout.

I'm not know if its the main cause, but I revised all my code, all $destroy, $diggest and $apply cicles, all $watchersand $listeners and i Just found it as a reason to break my application.

References: http://www.dwmkerr.com/fixing-memory-leaks-in-angularjs-applications/

**Comment by cthrax** [comment]
Thanks for your thorough investigation, it looks like you're on to something there. Right now we're in between primary contributors, so if you can come up with a solution for this problem it would be more than welcome. I'd be happy to help in whatever way I can, but my time is extremely limited right now.

**Comment by LeonanCarvalho** [comment]
Thanks for your answer @cthrax , could you explain about this function, is it extremely necessary?

**Comment by cthrax** [comment]
I'm not super familiar with this section of code, but it appears to be a failsafe function for multiple promises being created and to only resolve these promises once the unnecessary promises are resolved. Given that, I think this is necessary functionality, but it might be able to be implemented differently where the function is not needed. The functionality is necessary for two reasons, one is that the user is attaching to the same promises that the library is attaching to, requiring a degree of control over the promise not achievable if the user were to get to the promise first. Second, is that multiple instances of maps on the page need to resolve independently, but resolve externally as one promise. If that makes sense.

**Comment by LeonanCarvalho** [comment]
I think that's is the problem, I'm trying to attach instances controlled by "my Map Service", I need it because maps should be destroyed and created dynamically. (It's like a map-picker-input in my forms app).
So, my instances is not incremented in ctr var, and I could not identify it externally.
Testing it I made some tests:

```javascript
angular.module('uiGmapgoogle-maps.directives.api.utils').service('uiGmapIsReady', [
    '$q', '$timeout', function($q, $timeout) {
      var ctr, promises, proms;
      proms = [];
      promises = function() {
        return $q.all(proms).then(function(e){
            //When all uiGmaps promises have been resolved the ctr var and promise array should be reset.
            console.log(e);
            return proms.length = 0;
        });
      };
      return {
        spawn: function() {
          var d;
          d = $q.defer();
          d._resolved = false;
          d.promise.then(function(e){
              d._resolved = true;
              d._state = e;
              proms[id] = d;
//              proms.splice(id,1); //Remove promise from array?
              return e;
          });
          // Instead to use an auto increment in ctr to index all promises you should use the "proms.push" return:
          //  its return the length of array, its means the index the length - 1
          var id = proms.push(d) - 1;
          return {
            instance: id,//Instance is the index inside prom array
            deferred: d
          };
        },
        promises: promises,
        instances: function() {
          return prom.length;
        },
        promise: function(expect) {
          var d;
          
          //The expected promse exist?
          if(null != expect && undefined == proms[expect]){
              //This is the leak cause
              // When ohCrap comes here he trying to loop itself for an undefined expect.
              return; 
          }
          d = $q.defer();
          $timeout(function(){
            if (expect == null) { // resolves all promises
                d.resolve(promises());
            } else {//Resolve only expected promise
                //this is the new crap!
                console.log(proms[expect]);
                d.resolve(proms[expect]);
            }
          });
          
          return d.promise;
        },
        reset: function() {
          return proms.length = 0;
        },
        decrement: function(id) {
//          if (proms.length > 0 && proms.length <= id) {
//            return proms.splice(id,1);
//          } else {
            return proms.length = 0;
//          }
        }
      };
    }
  ]);
```

I could not completely solve the problem, also I did not dare to go so sink due to deploy stage, so I decided to use a slightly different approach in my application, where the management of instances would be on account of a service to the party who did not use the uiGmapIsReady.


I'll keep watching about that issue, if I have some divine revelation about that I return to share with you.

Thanks a lot for your support.

**Comment by cthrax** [comment]
Well thanks for your effort, hopefully we'll be able to address this at some point.

**Comment by nmccready** [comment]
I haven't been scanning the issues for a while. But this looks severe and interesting I can take a stab at it.

**Comment by nmccready** [comment]
One thing that would save memory right away is to move that internal function out so it is only created once.

**Comment by nmccready** [comment]
@LeonanCarvalho do you have plnkr which shows this happening? Does it show in the main example ?
Trying to use the Chrome Timeline to replicate what your have.

**Comment by nmccready** [comment]
Pretty sure I named it ohCrap because I knew it would blow up if it didn't find the expected app count. Besides adding time to the timeout how should this be handled to poll the ctr so we have the amount of maps we expect?

**Comment by nmccready** [comment]
Better idea for looping and keeping it async?

**Comment by LeonanCarvalho** [comment]
@nmccready Unfortunately I could not replicate it on Plunker at this moment because Its on a complex task with dynamics maps(a lot of them).
To manage when a Map is Ready I used this service to invoke maps.
As I comment on code below, the problem is with "ghost" instances that never come out, for exemple, when I have 3 maps and I invoke a 4th map, I expect map 4 but I just have 3 maps, so it's loops oh Crap function because I dont have the 4th map.

**Comment by nmccready** [comment]
Ok well that should be fixed now.

**Comment by nmccready** [comment]
PR merged this should be resolved on 2.0.X

**Comment by nmccready** [comment]
I'll cut another 2.0.X release
